### ISSUE TITLE
Load a component in a route depending on an asynchronous condition

### ISSUE THREAD

**Issue by alexjoverm** [angular/angular#12088]
**I'm submitting a ...**  (check one with "x")
- [ ] bug report => search github for a similar issue or PR before submitting
- [x] feature request
- [ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question

**Current behavior**
Doesn't seem possible to do it in an easy way without not use framework core stuff

**Expected behavior**
To be able to determine by an asynchronous condition (a condition that will be resolved later) which component to load in a route.

**What is the motivation / use case for changing the behavior?**
I'd say the following is a very common behavior expected on any medium size app:

Let's say you have a dashboard (just after you login). You wanna keep 2 different components UserDashboard and AdminDashboard (for not filling your template code with *ngIf), but that condition cannot be resolved immediately because the role comes from an API, so then you cannot "right away" conditionally pass one component or another to a specific route.

This doesn't seem to have an answer that implies not to compile and load a component by yourself.
Please, check out http://stackoverflow.com/questions/39688929/angular-2-how-to-conditionally-load-a-component-in-a-route-asynchronously 
- **Angular version:** 2.0.1

**Comment by DzmitryShylovich** [comment]
Use CanActivate guard and redirect to an appropriate page based on user role.

**Comment by alexjoverm** [comment]
But, doing that means you're redirecting to another route (let's say like "user-dashobard" or "admin-dashboard", while I wanted to keep them under "/dashboard".

Maybe it's just not the way it should be done? In that case, what are the drawbacks or the approach?

**Comment by pkozlowski-opensource** [comment]
> Maybe it's just not the way it should be done? In that case, what are the drawbacks or the approach?

IMO it would be super-confusing to have 2 completely different pages under the same URL. Imagine that someone who is an admin copies the `/dashboard` URL and sends it over to someone else that isn't. The other person would be able to access the page but would see completely different content from the intended one!

**Comment by DzmitryShylovich** [comment]
> But, doing that means you're redirecting to another route (let's say like "user-dashobard" or "admin-dashboard", while I wanted to keep them under "/dashboard".

I know it's a limitation but there're a lot of real bugs in the router and I don't think requested feature will be implemented in the nearest future.

**Comment by alexjoverm** [comment]
@pkozlowski-opensource didn't see it in that way, but it makes sense.

**Comment by amartone** [comment]
I've wanted this feature for some time, as well!

**Comment by bencameron** [comment]
I'm in need of this feature too.

For a few different applications we've written in Angular 1.x, we have a route that takes in an itemId as a parameter.  The item is then looked up by Id, and then the template/controller is selected based on the itemType.  We accomplished this in 1.x using ui-router's templateProvider and controllerProvider.

We will soon be upgrading these applications and need to figure out how to accomplish the equivalent in Angular 2.  Is there a better way to do this type of thing?

**Comment by HorusGoul** [comment]
Any update about this?

**Comment by thekalinga** [comment]
@vsavkin Is there any chance that this be implemented sometime soon?

**Comment by Misiu** [comment]
@bencameron I have exact same use case. Did You solve this?
Is this feature planned? Guards return true/false, maybe something similar that could return component?

**Comment by bencameron** [comment]
@Misiu we ended up just handling the dynamic selection of the component within another component that we route to.  For example, instead of routing to FooComponent or BarComponent based on some condition that we determine from the route params, we're routing to a component that we use as a container for the other components -- let's call it ContainerComponent.

Then, within the template for ContainerComponent, we're using a div with an ngSwitch directive on it to choose FooComponent or BarComponent based on the condition.

**Comment by Misiu** [comment]
@bencameron than's for reply. For now I'll stick with same solution.
@DzmitryShylovich any chances this could be added in near future?

**Comment by Misiu** [comment]
Is something like templateProvider and controllerProvider even possible in Angular (v4 or v5)?
ngSwitch is an option, but then You must add extra component that is just used for "routing".
Ideally this should be build in into core as a controlProvider option in route

**Comment by GermanKuber** [comment]
I have the same Problem. 
Someone have the good response about that now?

**Comment by Misiu** [comment]
@GermanKuber nope, using workaround as @bencameron suggested.
Sadly I see no comments from team members. I was hoping this could be added into version 6.0.

**Comment by simonKaplinski** [comment]
Yes, i have a good answer.

if i understand right, 
you want to navigate to specific route , and show another component, but not change the route.
not been redirected

you should use guard.
create guard (but don't worry, you won't be navigate to another route).

you'r route example: 
-----------------------------
{
 path:'...'
 component: '...',
 canActivate: [yourGuard]
}


your guard example:
-----------------------------
export clan yourGuard implements CanActivate{

canActivate(activatedRoute:ActivatedRouteSnapshot, state:RouteStateSnapshot):Observable{
  
  return this. canActivateNextRoute.switchMap(allowed => {
      if(allowed){
        return Observable.of(true)
      }
      
     //else....
     activatedRoute.routeConfig.component = ErrorMessageComponent (for example);
     activatedRoute.routeConfig.data = data (for example);

  })

  
  canActivateNextRoute():Observable<booloean>{
     return (do logic here...)
  }
 }
}

this solution work fine , if you get some difficult i can emplane more specific.

**Comment by klerick** [comment]
@simonKaplinski You solution is work only before init route. But if try to navigate some like `this.router.navigate(['/']);`
component does not change

**Comment by SerhiiKalaida** [comment]
This feature would be a great solution for **UI configuration**. Imagine we have an app with _multiple layout and style configurations_ to which we redirect a user based on database record. 

It's not a good option for the user to have routes like `'theme1/profile/:id'` and `'theme2/profile/:id'`. Instead we could conditionally switch to one of the components keeping the same router path:

   ```
 const routes: Routes = [
      {
        path: '',
        component: FooComponent,
        skippingCondition: [FooThemeActive],
      },
      {
        path: '',
        component: BarComponent,
        skippingCondition: [BarThemeActive],
      }
    ];
```

There are many questions about this problem on Stack Overflow without appropriate solutions:
[Angular2 - two components with the same route](https://stackoverflow.com/questions/37045548/angular2-two-components-with-the-same-route)
[How to bind two components from different lazy-loaded modules to the same route?
](https://stackoverflow.com/questions/49814235/how-to-bind-two-components-from-different-lazy-loaded-modules-to-the-same-route)
[Angular 2 routes with same URL?](https://stackoverflow.com/questions/34732245/angular-2-routes-with-same-url?noredirect=1&lq=1)

**Comment by Misiu** [comment]
@SerhiiKalaida what do You think about adding component factory.
So Your code could be simplified to:
```
const routes: Routes = [
   {
     path: '',
     component: FunctionThatReturnsDesiredComponentType()
   }
 ];
```

I'm not an expert but there is [ComponentFactoryResolver](https://angular.io/api/core/ComponentFactoryResolver) class. Maybe it could be used?

after looking at router config I've noticed that 
Route interface has component: Type < any > (https://github.com/angular/angular/blob/5.2.10/packages/router/src/config.ts#L347).
Would it be possible to add option to specify function that returns Type? @DzmitryShylovich what do You think? Is this even doable?

**Comment by Malinianne** [comment]
how to switch roles using get api in angular 5

**Comment by Misiu** [comment]
@pkozlowski-opensource any chance this might get added in near future? What do You think about [ComponentFactoryResolver ](https://angular.io/api/core/ComponentFactoryResolver)?

**Comment by Misiu** [comment]
@pkozlowski-opensource what about this? I'd like to give this a try, but I'm quite new to angular and I need a start point.
@IgorMinar what do You think about this feature request?
I have a very simple use case for this, but I'm not sure if this won't create issues with tree shaking, lazy loading, modules, Ivy and other mechanisms I'm not aware of.

**Comment by MattiJarvinen-BA** [comment]
I wonder if UrlMatcher could be used to achieve this as a workaround.

**Comment by MattiJarvinen-BA** [comment]
@Misiu 
I was thinking about using UrlMatchers this way:

```
const routes: Routes = [
  { 
    path: '',
    resolve: { loggedIn: false },
    children: [
      {
        matcher: loggedOutMatcher,
        redirectTo: '/login'
      },
      {
        matcher: loggedInMatcher,
        component: DashBoard
      },
    ]
];
```

**Comment by Misiu** [comment]
@MattiJarvinen-BA  so if I get this right You should be able to do this:

```
const routes: Routes = [
  { 
    path: 'detail/:id',
    resolve: { type: 'bug'},
    children: [
      {
        matcher: bugMatcher,
        component: BugDetailComponent
      },
      {
        matcher: featureRequestMatcher,
        component: FeatureRequestDetailComponent
      },
    ]
];
```

**Comment by MattiJarvinen-BA** [comment]
@Misiu unfortunately it seems that UrlMatchers do not wait for resolve.

**Comment by Misiu** [comment]
@MattiJarvinen-BA it looks like we have to wait for an official solution. Hopefully, it will get added soon.

**Comment by MattiJarvinen-BA** [comment]
@Misiu for my case I was able to add a matchers which check local storage for JWT token and depending on that matches the route.

Dashboard is in url '' for logged in users only and is lazy loaded. That module contains modules for logged in layout components ( menu & such ) + dashboard ( charts library ) so it's quite heavy and this was the only way to not load those before login when user accessed '' url since canLoad false in '' would prevent any further route checks including login.

```ts
export function matchLoggedInRoot(
  segments: UrlSegment[],
  group: UrlSegmentGroup,
  route: Route
) {

  let ret = null;

  // SSR future proofing
  // We do this harshly
  // - later CanActivate will redirect to login if token data isn't valid
  // - We cannot use services here just plain function
  if ( localStorage && localStorage.getItem && localStorage.getItem('tokenData') ) {
    ret = {consumed: []};
  }

  return ret;
}
```

```ts
export function matchNotLoggedInRoot(
  segments: UrlSegment[],
  group: UrlSegmentGroup,
  route: Route
) {
  let ret = null;
  if ( segments.length === 1 && segments[0].path === 'login') {
    ret = null;
  } else {
    // SSR future proofing
    // We do this harshly
    // - later CanActivate will redirect to login if token data isn't valid
    // - We cannot use services here just plain function
    if ( localStorage && localStorage.getItem && localStorage.getItem('tokenData') ) {
      ret = null;
    } else {
      // Check if we tried to match a real existing login requiring url
      // kinda...
      const loggedInFeatures = [
        'user',
        'page',
        'admin'
      ];
      if ( segments.length > 0 && loggedInFeatures.indexOf(segments[0].path) >= 0 ) {
          ret = {consumed: segments};

          if ( window && localStorage && localStorage.setItem ) {
            // set path to redirect after successful login
            // can't use route.navigate since we can't use services here
            // can't use route 
            localStorage.setItem('redirectUrl',
              window.location.pathname +
              window.location.hash +
              window.location.search);
          }
      } else if (segments.length === 0) {
        ret = {consumed: []};
      }
    }
  }

  return ret;
}
```

Instead of using localStorage for passing login redirect I could have used history.pushState maybe.

**Comment by nasreddineskandrani** [comment]
@pkozlowski-opensource 
I have a real use case where 2 type of users with a really really different `app-layout-component`.

So to avoid `ngIf` and also lazyload the needed content in one case and not the other via guard (with a double conditionnal in the route and the template) => this proposal become needed.

**Comment by imhuytq** [comment]
+1 for this feature.

**Comment by klerick** [comment]
+1 
I would like to have feature such as controllerProvider from angularjs ui router

**Comment by italomlp** [comment]
+1 for this.
I was searching about this feature in the Google and got here. As suggested by someone, I will use a central component (like Home) that will render my aproppriate component (MasterHome or AdminHome) by user type. It's not a good solution, 'cause I want to do it in the route config (less code for maintenance, etc.). Waiting for this feature.

**Comment by rocodesign** [comment]
I implemented something, it's still work in progress as there are a few bugs that need ironing out but the basic concept work, 

I need to implement vanityname of a project I'm working on, the url can lead to different types of entities as well as different Components, of course.

These are actual Modules as they have inner routing and different inner components, and also lazy loaded.

This is most important part of my ConditionalModule file

`
const moduleMap = {
  User: './containers/profile/profile.module#ProfileModule',
  Company: './containers/company/company.module#CompanyModule'
};

let name;
let routerSub;

let routerSub;

@Component({
  template: ''
})
export class EmptyComponent {}

@NgModule({
  imports: [RouterModule.forChild([{ path: '', component: EmptyComponent }])],
  providers: [ApiService],
  declarations: [EmptyComponent]
})
export class ConditionalModule {
  url;

  constructor(private router: Router, private api: ApiService) {
    if (routerSub) {
      routerSub.unsubscribe();
    }

    routerSub = router.events.subscribe(e => {
      const config = router.config;
      const last: any = config[0].children.find(r => r.path === ':name');

      if (e instanceof ActivationEnd) {
        let parentcount = 0;
        let parent = e.snapshot.parent;

        while (parent) {
          parentcount++;
          parent = parent.parent;
        }
       // only interested in routing events at this level
        if (parentcount !== 2) {
          return;
        }
       
        if (e.snapshot.params.name && name !== e.snapshot.params.name) {
          name = e.snapshot.params.name;
          api.getEntityName(name).subscribe(data => {
            const entityType = data.friendlyTypeName;
            const modulePath = moduleMap[entityType];
            const currentModule = get(last, '_loadedConfig.routes[0]');
            if (currentModule.loadChildren !== modulePath) {
              delete currentModule._loadedConfig;
              delete currentModule.component;
              currentModule.loadChildren = moduleMap[entityType];
              router.resetConfig(config);
              router.navigateByUrl('/' + name);
            }
          });
        }
      }
    });
  }`

while in the routing I added this as the last route. frankly I'm thinking about moving the conditions into a CanActivate, Resolver or urlMatcher, but for now I need to fix the remaining issues, as deep linking inside my modules and some better conditions to .
`{
        path: ':name',
        loadChildren: './containers/entity/conditional.module#ConditionalModule',
       data: {
          routeId: 'EntityRoute'
        }
      }`



Of course this solution can break at any point when angular is upgraded as it's not using `official api`.

**Comment by sic-sic** [comment]
+1 for this feature

**Comment by williangd** [comment]
+1 for this feature

**Comment by AlphaJuliettOmega** [comment]
+1 pretty please I have asynchronous events and I want to defer loading certain components until specific waves of async data calls back

**Comment by pickfire** [comment]
I thought routing should work the way like in express but it took me some time to find this issue.

**Comment by MarcusBondezan** [comment]
+1 for this feature

**Comment by KumarSidharth** [comment]
To some extent implementing route resolver can solve the problem.
[https://angular.io/guide/router#resolve-pre-fetching-component-data](url)

**Comment by Misiu** [comment]
@KumarSidharth an example would be really helpful.
I'm aware we can pre-fetch component data, but I want to display different component based on data that gets resolved.

**Comment by ugurcagridogan** [comment]
+1 for this feature

**Comment by seanzamora** [comment]
+1 for this feature

**Comment by seanzamora** [comment]
If there were multiple routes with the same path the RouterModule a next() method similar to express could be implemented that would simply go to the next route matching the path in question. I believe this could also be very useful and help us to keep our URL's clean. +1 for this feature!

`const routes: Routes = [
  {
    path: "",
    component: Templates.Login,
    children: [{ path: "forgot", component: Templates.Login }]
  },
  {
    path: "dashboard",
    loadChildren: () =>
      import("./modules/dashboard/admin/admin.module").then(m => m.AdminModule),
    canLoad: [CanLoadAdmin]
  },
  {
    path: "dashboard",
    loadChildren: () =>
      import("./modules/dashboard/employee/employee.module").then(
        m => m.EmployeeModule
      ),
    canLoad: [CanLoadEmployee]
  },
  {
    path: "dashboard",
    loadChildren: () =>
      import("./modules/dashboard/customer/customer.module").then(
        m => m.CustomerModule
      ),
    canLoad: [CanLoadImporter]
  }
];`

**Comment by BenRacicot** [comment]
@seanzamora that is exactly how [I expected the router to work](https://stackoverflow.com/questions/53937128/conditionally-load-module-on-the-empty-path-in-angular-router) 1.5 years ago. 
It's disappointing that a path cannot conditionally load a module. 
You MUST use redirection to a new path.

I truly hope the team adds this in the future.

**Comment by nasreddineskandrani** [comment]
@seanzamora your second solution is great for what we have now available.

But for the first proposal i don't like to repeat a route 3 times.
I see it this way:
```
{
    path: "dashboard",
    loadChildren: [
      { 
        module: () =
          import("./modules/dashboard/admin/admin.module").then(
              m = m.AdminModule
          ),
        canLoad: [CanLoadAdmin]
      },
      { 
        module: () =
          import("./modules/dashboard/client/client.module").then(
              m = m.ClientModule
          ),
        canLoad: [CanLoadClient]
      },
      ...
    ]
    }

```

**Comment by challamzinniagroup** [comment]
@seanzamora Your second solution is the best I've found for this - and demonstrates a great use case (and need) for this. Ultimately, you're doing with lazy-loaded modules what the router should allow for basic components. Fortunately in my case I'm dealing with lazy-loaded modules as well - so this worked like a charm. Thanks!

**Comment by nithipriya** [comment]
+1, how can we load 2 different component for same path even without lazy loaded modules

**Comment by KumarSidharth** [comment]
> +1, how can we load 2 different component for same path even without lazy loaded modules

Use auxiliary routes. Basically, you need to have two router outlets and to identify them you have to give them a name.

**Comment by nithipriya** [comment]
> > +1, how can we load 2 different component for same path even without lazy loaded modules
> 
> Use auxiliary routes. Basically, you need to have two router outlets and to identify them you have to give them a name.

Using auxiliary routes loads both the component at the same time in single page.
I want either of one component to be loaded conditionally for same path. Any idea?

**Comment by somcoder** [comment]
@nithipriya So you have have `OneComponent` and `AnotherComponent` and you want both of them to be served at `some-path` but only one of them to be shown conditionally, right? Create another component, say `WrapperComponent`, and point the `some-path` route to that wrapper component. Now, in the template of the `WrapperComponent` add the condition, something like this:
```html
<one *ngIf="condition"></one>
<another *ngIf="!condition"></another>
```
Hope that helps. 🙂

**Comment by nithipriya** [comment]
@somcoder Doing the same right now but was not sure if it's the right approch.
Thanks for confirming, this helps. :)

**Comment by nithipriya** [comment]
@somcoder Is there any other way to achieve this?

**Comment by somcoder** [comment]
@nithipriya Not quite sure really, but there could be other approaches as well.

**Comment by BenRacicot** [comment]
@nithipriya I had the same need and it can be done now with Ivy as components are rendered more independently than previous versions. The problem is that when you "lazy load" a component you run into trouble because ANY dependency has not been compiled with it. That component has been rendered on its own...

The solution was to create a feature module and lazy load THAT when you need to. Check out what I ended up with here:
[stackblitz.com/edit/lazy-module-wout-router](https://stackblitz.com/edit/lazy-module-wout-router?file=src%2Fapp%2Fapp.component.ts)

**Comment by nasreddineskandrani** [comment]
@BenRacicot **my solution is not far from yours. Check solution 2 full working also**
https://medium.com/front-end-tricks/dynamic-header-bars-and-or-contents-for-your-ivy-angular-app-56e1e7e3fc22

**Comment by BenRacicot** [comment]
@nasreddineskandrani wow! That is very similar, nice article BTW. We must be headed in the right direction :)

**Comment by Pingear** [comment]
**app-routing.module.ts ->**

```
export const routes: Routes = [
  {
    matcher: (url: UrlSegment[]) => {
      const persistanceService = new PersistanceService();
      return url.length === 1
        && url[0].path === 'home'
        && persistanceService.get(PersistanceKeys.CURENT_USER_TYPE) === UserType.CUSTOMER
        ? ({ consumed: url }) : null
    },
    canLoad: [CanCustomerLoadGuard],
    loadChildren: () => import('./customer/home/home.module').then(m => m.HomeModule)
  },
  {
    matcher: (url: UrlSegment[]) => {
      const persistanceService = new PersistanceService();
      return url.length === 1
        && url[0].path === 'home'
        && persistanceService.get(PersistanceKeys.CURENT_USER_TYPE) === UserType.EMPLOYEE
        ? ({ consumed: url }) : null
    },
    canLoad: [CanEmployeeLoadGuard],
    loadChildren: () => import('./employee/home/home.module').then(m => m.HomeModule)
  },
]

@NgModule({
  imports: [ RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
```
**can-employee-load.guard.ts ->**

```
import { Injectable } from "@angular/core";
import { CanLoad, Route, UrlSegment, UrlTree } from "@angular/router";
import { select, Store } from "@ngrx/store";
import { Observable } from "rxjs";
import { map, take } from "rxjs/operators";
import { currentUserSelector } from 'src/app/auth/store/auth.selector';
import { ConsructorNames } from '../entities/enums';

@Injectable()
export class CanEmployeeLoadGuard implements CanLoad {
    constructor(private _authService: AuthService) { }
    canLoad(route: Route, segments: UrlSegment[]): boolean | UrlTree | Observable<boolean | UrlTree> | Promise<boolean | UrlTree> {
        return this._authService.user.pipe(
            take(1),
            map(user => {
                if (user) return user.constructor.name === ConsructorNames.EMPLOYEE;
                else return false;
            })
        );
    }
}
```

**somewhere onLogin() ->**

```
currentUser = new Employee(userDto);
this.persistanceService.set(PersistanceKeys.CURENT_USER_TYPE, UserType.EMPLOYEE);

```
**persistance.service.ts ->**

```
import { Injectable } from '@angular/core';

@Injectable()
export class PersistanceService {

  constructor() { }

  set(key: string, data: any): void {
    try {
      localStorage.setItem(key, JSON.stringify(data));
    } catch (error) {
      console.error('Error saving to lacalStorage => ', error);
    }
  }
  get(key: string): any {
    try {
      return JSON.parse(localStorage.getItem(key));
    } catch (error) {
      console.error('Error getting data from lacalStorage => ', error);
      return null;
    }
  }
}
```


**enums.ts ->**

```
export enum UserType {
    CUSTOMER = 'customer',
    EMPLOYEE = 'employee'
}

export enum PersistanceKeys {
    CURENT_USER_TYPE = "userType"
}

export enum ConsructorNames {
    CUSTOMER = "Customer",
    EMPLOYEE = "Employee"
}
```

**Comment by mightymatth** [comment]
For hours and hours, I'm desperately trying to implement an optional segment in my routing, depending on an async backend call, but I just can't believe that **4.5 years** and **several related issues**, this is still not possible.

Let me explain my intents, which are somewhat different from some mentioned above, but they are in the same category.
Imagine having a  multi-tenant application for restaurants with the following routes:

```
https://multi-restaurants.com/:restaurantName/menu
https://multi-restaurants.com/:restaurantName/menu/:id
https://multi-restaurants.com/:restaurantName/settings
...
```

OK, everything looks fine. But, what about having another single domain pointing to a single restaurant? In other words, we'd like to have something like this:

```
https://special-restaurant.com/menu
https://special-restaurant.com/menu/:id
https://special-restaurant.com/settings
...
```

Pay attention to what I want to achieve.
While routing, check if `window.hostname` is equal to the registered restaurant's hostname (backend API call). If true, load children for the entire application (let's say we call it `appRoutes`). If false, route to the route with path `:restaurantName` which will have a guard to check if the restaurant with this name exists; if true, load children with `appRoutes`; if false, navigate to default restaurant.

To achieve that, as I see with previous replies, we need to do nasty sorceries. 

Nice-to-have options:
#### Custom matcher
As @MattiJarvinen and @Misiu tried the solution [here](https://github.com/angular/angular/issues/12088#issuecomment-436916696) – a great idea, but injecting a service and wait for async calls is hackery and I'm not sure it's even possible. If a custom matcher could use DI and return `Observable<boolean>`, it would be nice.

#### next() method
As @seanzamora proposed the idea [here](https://github.com/angular/angular/issues/12088#issuecomment-591760284), this would be perfect if we could call `.next()` method somewhere in route guard so we can continue with matching the next route.

So, if I have:

```
const routes: Routes = [
    {path: '', canActivate: [CustomHostnameGuard], children: appRoutes},
    {path: ':restaurantName', canActivate: [RestaurantGuard], children: appRoutes},
];
```
This way, it will always match the first empty path, and if we decide not to match the route based on API call, it would be nice to have `.next()` method on the router to continue matching the next route, in this case `:restaurantName` route. 

--- 

All other hackeries with `canLoad` and lazy-loaded modules seem too complicated to achieve this functionality.

**Comment by nasreddineskandrani** [comment]
@mightymatth 

**CASE: https://multi-restaurants.com/:restaurantName/menu**
https://angular-ivy-multi.stackblitz.io/ngfood/menu

**CASE: https://angular-ivy-specific.stackblitz.io/menu**
https://specific-restaurant.com/menu

code: https://stackblitz.com/edit/angular-ivy-multi?file=src%2Fapp%2Fapp.module.ts
note: 
1. you can change the name of the project in stackblitz  (fork and remove multi from name to see the specific restaurent one)
2. update the url with `/ngfood/menu` or `/menu` 
3. `appRoutes` is not common (i didn't do it but you can just do common module etc...)
4. `checkIfRestaurentExist` can access storage/window variable (set value in canActivate)  OR  split the pathname to get the restuarent name to do your backend call

Hope this helps.

**Comment by mightymatth** [comment]
@nasreddineskandrani, thank you for your time! I see that this approach work, but we can agree that it seems overly complex for such an easy task. 
When I try to solve some problem, I prefer not to chose the first working approach, no matter how ugly and nasty it is. I always ask myself: "Is this simple enough to even try it? Is there an easier way? When somebody comes after me, would anyone get it why it's that complicated or there was an easier way?"

I said earlier that I don't like this approach because I need to create two separate modules which contain all the same routes and components, but one has this `path: ':restaurantName'` route. Besides that, they are identical.

For me, at the moment, it's not worth complicating the code if nobody else after me would understand this nasty workaround with a big commend above all: "Don't touch it, it works!"

Thank you once again for working examples. However, we must seek a better, easier way of doing matching asynchronous path matching.

**Comment by nasreddineskandrani** [comment]
> @nasreddineskandrani, thank you for your time! I see that this approach work, but we can agree that it seems overly complex for such an easy task.
> When I try to solve some problem, I prefer not to chose the first working approach, no matter how ugly and nasty it is.
When you have a client you need to provide a solution. You can wait for the feature request in angular to be processed.
That's why I tried to solve it. I let the angular team decide what they want to add to address this. For me it remains an acceptable solution for now.


> I said earlier that I don't like this approach because I need to create two separate modules which contain all the same routes and components, but one has this path: ':restaurantName' route. Besides that, they are identical.
I processed my note number 4 to demonstrate how to remove the duplication since it was a quick draft I did yesterday.
find the updated code here: https://stackblitz.com/edit/angular-ivy-multi-updated?file=src/app/site/site-routes.ts
Again this shows a working solution not saying it's the best.

**Comment by rohittidke** [comment]
facing a similiar scenario, other than "not pretty" workarounds I don't see any other solution to the problems.
my scenario -> based on the selection company -> the routing tree differs slightly, and the same url can re-direct to two different components. 

It would be really cool if the Angular team comes up with a feature to solve such use-cases.

**Comment by rohittidke** [comment]
I found an intermediate solution if it helps anyone and it works very well and is performant. 

This is how the route looks like

`  {
            path: 'account',
            data: {
              breadcrumb: 'My Account'
            },
            children: [
              {
                path: 'my-account-status', component: InterfaceIdBasedRouting,
                data: {
                  breadcrumb: 'My Account Status',
                  keys: {
                    60: MyAccountStatusComponent,
                    64: OtherMyAccountStatusComponent
                  }
                }
              },`

The InterfaceIdBasedRouting looks like this

`@Component({
  selector: 'app-conditional-router',
  templateUrl: './linkid-router.component.html'
})
export class InterfaceIdBasedRouting implements AfterViewInit, OnInit {

  @ViewChild(PlaceholderDirective) placeholder: PlaceholderDirective;

  componentInitialized$ = new BehaviorSubject<boolean>(false);

  constructor(private ComponentFactoryResolver: ComponentFactoryResolver, private someAsyncServiceYouDependUpon: SomeAsyncServiceYouDependUpon, private routerEventService: RouterEventService, private activatedRoute: ActivatedRoute) { }

  ngAfterViewInit(): void {

    combineLatest([this.routerEventService.currentUrl$, this.someAsyncServiceYouDependUpon.someAsyncResult$, this.componentInitialized$])
      .pipe(
        map(([a, b, c]) => ({ url: a, someAsyncResult: b, initialized: c })),
        switchMap(e => {
          if (e.initialized) {
            const returnNumberInHashTable= e.someAsyncResult.returnNumberInHashTable;
            const routeData = this.activatedRoute.snapshot.data;
            if (navItem?.linkId) {
              if (returnNumberInHashTable in routeData.keys) {
                return of(returnNumberInHashTable); // this structure is required to run this component only once (later you see takeWhile)
              }
            }
          }
          return of(undefined);
        }),
        takeWhile(linkId => linkId !== undefined)
      ).subscribe(linkId => this.createComponent(linkId));
  }

  ngOnInit(): void {
    this.componentInitialized$.next(true);
  }

  createComponent(linkId: number) {
    this.placeholder.ViewContainerRef.clear();
    const routeData = this.activatedRoute.snapshot.data;
    const componentFactory = this.ComponentFactoryResolver.resolveComponentFactory(routeData.keys[linkId]);
    const componentRef = this.placeholder.ViewContainerRef.createComponent(componentFactory);
  }
}`

The placeholder directive

`import { Directive, ViewContainerRef } from '@angular/core';

@Directive({
  selector: '[appPlaceholder]'
})
export class PlaceholderDirective {
  constructor(public ViewContainerRef: ViewContainerRef) {

  }
}`

**Comment by ArtemGrachov** [comment]
After hours of search, I have found this article: https://www.linkedin.com/pulse/how-make-route-load-different-components-based-async-veloso-alves/ with very simple and clear solution.
